### ISSUE TITLE
Optimizing Kafka Data Consumption and Join Performance in ClickHouse

### ISSUE THREAD

**Issue by aol-59** [ClickHouse/ClickHouse#56925]
<img width="718" alt="Join" src="https://github.com/ClickHouse/ClickHouse/assets/50527487/9bc87c05-8bbd-4162-8acd-88cd023ce5ce">

**Clickhouse Version : 23.8.1.2992**

I'm working with a Kafka setup where I have **16 million** records distributed across 20 partitions. Notably, the ClickHouse Kafka engine is consuming data at a rate of **300K** records every **5 seconds**. I've set up a series of processes and tables in ClickHouse as follows:

1- Created a Kafka engine table named '**CH_Kafka_engine_queue_consumer**' to consume data from a Kafka topic.

2- Set up a Materialized View (MV) named '**CH_MV_consume**' to process data from this topic.

3- Created a target table named '**CH_kafka_target**' to store the consumed data.

4- Established another MV to read data from '**CH_kafka_target**', perform a join operation with a '**lookup_table**' containing **88K** duplicate records, **not all are duplicates**, and then insert the joined data into a table called '**CH_JOIN_target**'.

5- Finally, I'm pushing data back to Kafka using a Kafka engine and MV.

Here's the SQL query I'm using for the join operation:

`SELECT columns
FROM CH_kafka_target
LEFT JOIN (
    SELECT columns, MAX(date)
    FROM lookup_table
    GROUP BY columns
) ON (id = id)
`

**This query is designed to group data based on the maximum date to eliminate duplicates.**

i removed the **group by** and nothing changed in performance.

Currently, the time to consume data alone is 2 minutes, but when I include the data consumption and join operation, it increases to 15 minutes.

I've tried adjusting settings in 'CH_MV_JOIN' based on ClickHouse documentation suggestions:

-  Initially: Settings **max_threads = 2, join_algorithm = 'hash'**
- Then changed to: Settings **max_threads = 20, join_algorithm = 'hash'**
- Finally tried: Settings **max_threads = 20, join_algorithm = 'parallel_hash'**
-
`SELECT columns
FROM CH_kafka_target
LEFT JOIN (
    SELECT columns, MAX(date)
    FROM lookup_table
    GROUP BY columns
) ON (id = id)
Settings max_threads = 20, join_algorithm = 'parallel_hash';
`

Despite these changes, there hasn't been any significant improvement in performance.

**When I run the query manually to join all 16M with 88K lookup table and change the settings the performance increases and returns the result in 1~2 seconds but nothing changes in MV.**

 Can anyone offer advice or insights on how to optimize this process?

**Comment by den-crane** [comment]
Just add a subquery to limit the right table

```sql
SELECT columns FROM CH_kafka_target 
   LEFT JOIN ( SELECT columns, MAX(date) 
               FROM lookup_table 
               where id in (select id from CH_kafka_target)                  -----<<<<<<<<<-------
               GROUP BY columns 
              ) ON (id = id)
```

**Comment by den-crane** [comment]
CH_kafka_target, CH_MV_consume look excessive, why do you need them?

**Comment by aol-59** [comment]
> CH_kafka_target, CH_MV_consume look excessive, why do you need them?

I need it for testing purposes only, to calculate the difference in insertion time at each step. I removed it, but nothing changed. Can I add 'Settings' to the Materialized View? I tried adding it, but there was no change in performance. As I mentioned before, when I manually ran the query on all 16 million records, the results came back in seconds.
